### PR TITLE
Cosmos DB: Rewrite RowConverter

### LINKED ISSUE #18150: Transport ObjectNode to Spark Row

### LINKED ISSUE #18151: Transform Spark Internal Row to ObjectNode

### PR DESCRIPTION
**Pull request by ealsur**
This PR rewrites the CosmosRowConverter completely, adding support for:

* InternalRow -> ObjectNode
* Row -> ObjectNode
* ObjectNode -> Row
* ObjectNode -> InternalRow

And adding new unit tests (Spec) to cover the behavior on the converter.

Closes #18151
Closes #18150

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala [hunk 1]
```diff
@@ -3,246 +3,220 @@
...
-  // TODO: moderakh this method requires a rewrite
-  // this is borrowed from old OLTP spark connector
-  def documentToMap(document: ObjectNode): Map[String, AnyRef] = {
-    if (document == null) {
-      new HashMap[String, AnyRef]
-    } else {
-      getMap(document).asScala.toMap
+    extends CosmosLoggingTrait {
+
+    val objectMapper = new ObjectMapper()
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala:25]: one of the issues we ran into with older oltp spark connector was duplicate json fields which couldn't be fetched from cosmos db by spark.

handling such scenarios require configuring how ObjectMapper and conversion works.

Please add a TODO here capturing that we need to make this behaviour configure or cover similar scenarios.

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala [hunk 2]
```diff
@@ -3,246 +3,220 @@
...
-    row.schema.fields.zipWithIndex.foreach({
-      case (field, i) => {
-        objectNode.set(field.name, convertToJson(row.get(i), field.dataType, isInternalRow = false))
+
+    def fromRowToObjectNode(row: Row): ObjectNode = {
+        val objectNode: ObjectNode = objectMapper.createObjectNode()
+        row.schema.fields.zipWithIndex.foreach({
+            case (field, i) =>
+                field.dataType match {
+                    case NullType  | _ if row.isNullAt(i) => objectNode.putNull(field.name)
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala:42]: can be added later after this PR. please make sure we add test coverage for null case in both conversion directions

**Review by moderakh — APPROVED** [review]
awesome work. @ealsur 

I see that you already added many new unit tests. that's great thanks.
Please make sure we add test coverage for every single type (nested, array, null, etc)

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala (hunk 1 above)
Comment (ealsur) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala:27]: Added TODO

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala (hunk 2 above)
Comment (ealsur) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala]: Added a new test for that explicit scenario called:

`"null in spark row" should "translate to ObjectNode"`

**Comment by ealsur** [comment]
> awesome work. @ealsur
> 
> I see that you already added many new unit tests. that's great thanks.
> Please make sure we add test coverage for every single type (nested, array, null, etc)

Thanks @moderakh ! Current test suite covers basic types (strings, numbers), date, time, map, struct, null, array, nested in both directions (Row to Json and Json to Row).

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala [hunk 3]
```diff
@@ -3,246 +3,224 @@
...
+
+    // scalastyle:off
+    private def convertSparkDataTypeToJsonNode(fieldType: DataType, rowData: Any) : JsonNode = {
+        fieldType match {
+            case StringType if rowData.isInstanceOf[String] => objectMapper.convertValue(rowData.asInstanceOf[String], classOf[JsonNode])
+            case StringType if rowData.isInstanceOf[UTF8String] => objectMapper.convertValue(rowData.asInstanceOf[UTF8String].toString, classOf[JsonNode])
+            case BinaryType => objectMapper.convertValue(rowData.asInstanceOf[Array[Byte]], classOf[JsonNode])
+            case BooleanType => objectMapper.convertValue(rowData.asInstanceOf[Boolean], classOf[JsonNode])
+            case DoubleType => objectMapper.convertValue(rowData.asInstanceOf[Double], classOf[JsonNode])
+            case IntegerType => objectMapper.convertValue(rowData.asInstanceOf[Int], classOf[JsonNode])
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala:75]: These conversions have been part of heated discussion sometimes. Common challenges are:

Long, Double, Float, Decimal --> When converting them to json as number you always risk loosing precision - so customers sometimes need to convert it to string instead to preserve precisions - this of course makes it impossible to use any aggregates/numeric functions within Cosmos native queries.

DateType/TimestampType --? Contention whether to convert into Long (Epoch of seconds vs. Epoch of milliseconds) or into a string representation - like ISO 8601-2019

I am wondering whether we should make this configurable - and if so rather as a way of setting a global flag (preferTextualJsonRepresentation vs. prefernativeJsonRepresenttation which would be used for all applicable columns/data types - or whether this would need ot be configurable ona  per column level.
Ultimately tehre is always the workaround for custoemrs to do the conversion on the dta frame upstream/downstream - but I think this has been a contention point often enough to have a brief discussion to decide how we want to proceed. Makes sense?

**Review by FabianMeiswinkel — CHANGES_REQUESTED** [review]
Very nice job Matias! only one comment left - not necessarily blocking - but I want to have a brief discussion about it to decide how we want to proceed before merging.

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala (hunk 3 above)
Comment (ealsur) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala:81]: I added Json > Spark Row support for fields that are strings into DateType.Timestamp and DataType.Date by trying to parse the string as ISO 8601 with `+HH:mm` timezone notation or `Z` for UTC.

The Spark Row > Json serialization, that is something that might require extra configuration work and outside of the scope of this PR.

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala [hunk 4]
```diff
@@ -3,246 +3,224 @@
...
+            // Assuming the array is of bytes
+            objectMapper.convertValue(arrayNode, classOf[Array[Byte]])
+        case (_, _: BooleanType) => value.asBoolean()
+        case (_, _: StringType) => value.asText()
+        case (_, _: DateType) => toDate(value)
+        case (_, _: TimestampType) => toTimestamp(value)
+        case (isJsonNumber(), DoubleType) => value.asDouble()
+        case (isJsonNumber(), DecimalType()) => value.decimalValue()
+        case (isJsonNumber(), FloatType) => value.asDouble()
+        case (isJsonNumber(), LongType) => value.asLong()
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala:198]: For these I think we also need a similar translation of text --> Float, Text -->Double etc. like for Date and Timestamp

Comment (ealsur) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala:198]: Done 😄

**Review by FabianMeiswinkel — APPROVED** [review]
Thanks!
